api/v2/api.go

```diff
@@ -672,3 +672,31 @@ func parseFilter(filter []string) ([]*labels.Matcher, error) {
 	}
 	return matchers, nil
 }
+
+var (
+	swaggerSpecCacheMx sync.Mutex
+	swaggerSpecCache   *loads.Document
+)
+
+// getSwaggerSpec loads and caches the swagger spec. If a cached version already exists,
+// it returns the cached one. The reason why we cache it is because some downstream projects
+// (e.g. Grafana Mimir) creates many Alertmanager instances in the same process, so they would
+// incur in a significant memory penalty if we would reload the swagger spec each time.
+func getSwaggerSpec() (*loads.Document, error) {
+	swaggerSpecCacheMx.Lock()
+	defer swaggerSpecCacheMx.Unlock()
+
+	// Check if a cached version exists.
+	if swaggerSpecCache != nil {
+		return swaggerSpecCache, nil
+	}
+
+	// Load embedded swagger file.
+	swaggerSpec, err := loads.Analyzed(restapi.SwaggerJSON, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to load embedded swagger file: %v", err.Error())
```

Comment:
Should we not keep the original `err`?

```suggestion
		return nil, fmt.Errorf("failed to load embedded swagger file: %v", err)
```

Reply:
I just kept the previous error. If we want to switch it to error wrapping, then it should be:

```suggestion
		return nil, fmt.Errorf("failed to load embedded swagger file: %w", err)
```

(notice the `%w`).

Reply:
Perhaps the original code was written before `%w` was added to `fmt`. I think it would be nice to change it to `%w`, unless there is a good reason against it! :slightly_smiling_face:

Reply:
Agree, let me do it.